yang/frr-zebra.yang

```diff
@@ -0,0 +1,1887 @@
+	description
+	  "ICMP admin-prohibited.";
+      }
+    }
+    default "null";
+    description
+      "Nexthop blackhole types.";
+  }
+
+  typedef gateway-address {
+    type union {
+      type inet:ipv4-address;
+      type inet:ipv6-address;
+    }
+  }
+
+  /*
+   * Nexthop object
+   */
+  grouping frr-nexthop {
+    leaf nh-type {
+      type frr-zebra:nexthop-type;
+      mandatory true;
+      description
+	"The nexthop type.";
+    }
+
+    leaf gateway {
+      type frr-zebra:gateway-address;
+      description
+	"The nexthop gateway address.";
+    }
+
+    leaf vrf {
+      type string;
+      description
+	"The nexthop vrf name, if different from the route.";
+    }
+
+    leaf interface {
+      type string;
+      description
+	"The nexthop egress interface.";
+    }
+
+    leaf bh-type {
+      type frr-zebra:blackhole-type;
+      description
+	"A blackhole sub-type, if the nexthop is a blackhole type.";
+    }
+
+    leaf flags {
+      type uint32;
+      description
+	"The nexthop's raw flags value.";
+    }
+
+    leaf is-duplicate {
+      type empty;
+      description
+	"Duplicate nexthop; will be ignored.";
+    }
+    leaf is-recursive {
+      type empty;
+      description
+	"Nexthop must be resolved through another gateway.";
+    }
+    leaf is-onlink {
+      type empty;
+      description
+	"Nexthop is directly connected.";
+    }
+    leaf is-active {
+      type empty;
+      description
+	"Nexthop is active.";
+    }
+
+    uses rt-types:mpls-label-stack {
+      description
+	"Nexthop's MPLS label stack.";
+    }
+
+    leaf mtu {
+      type uint32;
+      description
+	"The nexthop's specific MTU.";
+    }
+
+  } // End of nexthop
+
+  /*
+   * Nexthop-group container
+   */
+  grouping frr-nexthop-group {
+    description
+      "A nexthop-group, represented as a list of nexthop objects.";
+
+    leaf name {
+      type string;
+      description
+        "The nexthop-group name.";
+    }
+
+    list entry {
+      key "id";
+      description
+        "A list of nexthop objects.";
+      leaf id {
+        type uint32;
+        description
+          "Identifies a nexthop within a nexthop group; the entries
+           are ordered by id value, and the value has no other meaning.";
+      }
+
+      uses frr-zebra:frr-nexthop;
+
+    }
+  } // End of frr-nexthop-group
+
+  /*
+   * Common route data, shared by v4 and v6 routes.
+   */
+  grouping route-common {
+    description
+      "Common information about a route.";
+
+    leaf vrf {
+      type string;
+      description
+	"The route's vrf name.";
+    }
+
+    leaf distance {
+      type uint32;
+      description
+        "Admin distance based on routing protocol.";
+    }
+    leaf metric {
+      type uint32;
+      description
+        "Route metric value.";
+    }
+    leaf tag {
+      type uint32 {
+        range "1..4294967295";
+      }
+      description
+        "Route tag value.";
+    }
+
+    leaf is-selected {
+      type empty;
```

Comment:
Wouldn't it be better to use `boolean` type instead of `empty`?

The `boolean` you can check for `true`/`false`, `empty` on the other hand you have to check if it exists or not (at least this is what I understood from this type).

Same applies to other `empty` instances in this node.

Reply:
The use of `empty` was suggested by some of our yang experts. To me, anyway, presence is "true", and absence is "false", and that seems pretty clear. There's also a certain compact-ness that I sort of like: we don't have to send around a lot of "false" attributes - the false ones are just not present.
I wanted to put the topic out there, see what the code (model) would look like if I took this path. I agree that it would be good to have some consistent style guidelines for frr models.

Reply:
I tend to think "empty" standing in for "false" is okay here -- though receivers should really be able to handle both possibilities, maybe?